Refactors and simplifies metadata cache, moves ignore access into session

mrclay opened the pull request:
This does several things:
- Moves "ignore access" API into ElggSession so session becomes the single source of access visibility in the system. Tries not to break old code like `elgg_get_access_object()->set_ignore_access()`
- Makes the MD cache depend on session so it can properly segregate data by visibility
- Optimizes the cache for the most common case of read-only access. We no longer keep track of whether individual MD names are synchronized with the DB, nor do we try to inject MD writes into the cache. This greatly simplifies the hard task of cache invalidation, which was a good hiding place for bugs.
- MD cache tests overhauled and improved.

Fixes #7953

mrclay committed:
chore(cache): simplifies metadata cache

This optimizes the cache for the most common case of read-only access. We no
longer keep track of whether individual MD names are synchronized with the DB,
nor do we try to inject MD writes into the cache. This greatly simplifies the
hard task of cache invalidation, which was a good hiding place for bugs.

This also segregates the cache by access visibility, which makes the cache
safer to use, and frees us from having to invalidate when ignore access is
toggled.

Fixes #7953

mrclay committed:
chore(session): moves ignore access state into session

mrclay commented on engine/classes/Elgg/Access.php:
```diff
@@ -1,74 +1,14 @@
-	 * @deprecated 1.8 Use \Elgg\Access::getIgnoreAccess()
-	 */
-	public function get_ignore_access() {
-		elgg_deprecated_notice('\Elgg\Access::get_ignore_access() is deprecated by \Elgg\Access::getIgnoreAccess()', 1.8);
```
We broke this by changing the class name in 1.10. Do we care about keeping this?

juho-jaakkola replied:
We're not exposing this to the public API, so should be safe to remove.

mrclay replied:
I chose to keep elgg_get_access_object() around. No real harm keeping it.

juho-jaakkola commented on engine/lib/deprecated-1.10.php:
```diff
@@ -36,7 +36,9 @@ function file_get_general_file_type($mime_type) {
 	elgg_deprecated_notice(__FUNCTION__ . ' is deprecated. Use elgg_get/set_ignore_access()', '1.10');
-	return _elgg_services()->access;
+
+	// note we've already broken this function (it used to return an ElggAccess)
```
elgg_get_access_object() is private, so this shouldn't matter

mrclay replied:
It was introduced in 1.7 and [marked private in 1.8](https://github.com/mrclay/Elgg-leaf/commit/cb5df9222c274a6fcf9263cc5fb905b7a76add40#diff-23234856e075522f9fb90e4cbc1d4ae1R906). I think I agree it can go.

mrclay committed:
chore(access): removes the Access service

This keeps code like elgg_get_access_object()->get_ignore_access() working
as long as the plugin doesn’t demand type ElggAccess or Elgg\Access. This
seems a reasonable compromise.

mrclay commented on engine/classes/ElggSession.php:
```diff
@@ -250,11 +254,67 @@ public function isLoggedIn() {
+	 * @deprecated 1.8 Use elgg_get_ignore_access()
+	 */
+	public function get_ignore_access() {
+		return $this->getIgnoreAccess();
```
These don't throw deprecation notices because they would only be accessed via elgg_get_access_object(), which does.

mrclay commented:
Pushing forward with this in hopes of tackling https://github.com/Elgg/Elgg/issues/4346#issuecomment-66770626
